File: rs/tests/test_apply.rs
mod common;

use gitstore::*;

// ---------------------------------------------------------------------------
// WriteEntry construction
// ---------------------------------------------------------------------------

#[test]
fn write_entry_from_bytes() {
    let e = WriteEntry::from_bytes(b"data".to_vec());
    assert_eq!(e.data.as_deref(), Some(b"data".as_ref()));
    assert!(e.target.is_none());
    assert_eq!(e.mode, MODE_BLOB);
}

#[test]
fn write_entry_from_text() {
    let e = WriteEntry::from_text("hello");
    assert_eq!(e.data.as_deref(), Some(b"hello".as_ref()));
    assert!(e.target.is_none());
    assert_eq!(e.mode, MODE_BLOB);
}

#[test]
fn write_entry_symlink() {
    let e = WriteEntry::symlink("target.txt");
    assert!(e.data.is_none());
    assert_eq!(e.target.as_deref(), Some("target.txt"));
    assert_eq!(e.mode, MODE_LINK);
}

// ---------------------------------------------------------------------------
// WriteEntry validation
// ---------------------------------------------------------------------------

#[test]
fn write_entry_validate_blob_ok() {
    let e = WriteEntry::from_bytes(b"data".to_vec());
    assert!(e.validate().is_ok());
}

#[test]
fn write_entry_validate_symlink_ok() {
    let e = WriteEntry::symlink("target");
    assert!(e.validate().is_ok());
}

#[test]
fn write_entry_validate_symlink_without_target() {
    let e = WriteEntry {
        data: None,
        target: None,
        mode: MODE_LINK,
    };
    assert!(e.validate().is_err());
}

#[test]
fn write_entry_validate_symlink_with_data() {
    let e = WriteEntry {
        data: Some(b"data".to_vec()),
        target: Some("target".into()),
        mode: MODE_LINK,
    };
    assert!(e.validate().is_err());
}

#[test]
fn write_entry_validate_blob_without_data() {
    let e = WriteEntry {
        data: None,
        target: None,
        mode: MODE_BLOB,
    };
    assert!(e.validate().is_err());
}

#[test]
fn write_entry_validate_blob_with_target() {
    let e = WriteEntry {
        data: Some(b"data".to_vec()),
        target: Some("target".into()),
        mode: MODE_BLOB,
    };
    assert!(e.validate().is_err());
}

#[test]
fn write_entry_validate_exec_ok() {
    let e = WriteEntry {
        data: Some(b"#!/bin/sh".to_vec()),
        target: None,
        mode: MODE_BLOB_EXEC,
    };
    assert!(e.validate().is_ok());
}

#[test]
fn write_entry_validate_unsupported_mode() {
    let e = WriteEntry {
        data: Some(b"data".to_vec()),
        target: None,
        mode: 0o777777,
    };
    assert!(e.validate().is_err());
}

// ---------------------------------------------------------------------------
// apply — basic writes
// ---------------------------------------------------------------------------

#[test]
fn apply_single_bytes() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.apply(
        &[("data.bin", WriteEntry::from_bytes(b"\x00\x01\x02".to_vec()))],
        &[],
        Default::default(),
    )
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read("data.bin").unwrap(), b"\x00\x01\x02");
}

#[test]
fn apply_single_text() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.apply(
        &[("hello.txt", WriteEntry::from_text("hello world"))],
        &[],
        Default::default(),
    )
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("hello.txt").unwrap(), "hello world");
}

#[cfg(unix)]
#[test]
fn apply_symlink_entry() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.apply(
        &[("link", WriteEntry::symlink("target.txt"))],
        &[],
        Default::default(),
    )
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.file_type("link").unwrap(), FileType::Link);
    assert_eq!(fs.readlink("link").unwrap(), "target.txt");
}

#[cfg(unix)]
#[test]
fn apply_executable_entry() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let entry = WriteEntry {
        data: Some(b"#!/bin/sh".to_vec()),
        target: None,
        mode: MODE_BLOB_EXEC,
    };
    fs.apply(&[("run.sh", entry)], &[], Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.file_type("run.sh").unwrap(), FileType::Executable);
    assert_eq!(fs.read("run.sh").unwrap(), b"#!/bin/sh");
}

// ---------------------------------------------------------------------------
// apply — multiple entries
// ---------------------------------------------------------------------------

#[test]
fn apply_multiple_entries_single_commit() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let hash_before = fs.commit_hash().unwrap();

    fs.apply(
        &[
            ("a.txt", WriteEntry::from_text("aaa")),
            ("b.txt", WriteEntry::from_text("bbb")),
            ("c.txt", WriteEntry::from_text("ccc")),
        ],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_ne!(fs.commit_hash().unwrap(), hash_before);
    assert_eq!(fs.read_text("a.txt").unwrap(), "aaa");
    assert_eq!(fs.read_text("b.txt").unwrap(), "bbb");
    assert_eq!(fs.read_text("c.txt").unwrap(), "ccc");

    // Single commit
    let log = fs.log(fs::LogOptions { limit: Some(5), ..Default::default() }).unwrap();
    assert_eq!(log.len(), 2); // init + apply
}

#[cfg(unix)]
#[test]
fn apply_mixed_types() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    let exec_entry = WriteEntry {
        data: Some(b"#!/bin/sh".to_vec()),
        target: None,
        mode: MODE_BLOB_EXEC,
    };
    fs.apply(
        &[
            ("file.txt", WriteEntry::from_bytes(b"data".to_vec())),
            ("link", WriteEntry::symlink("file.txt")),
            ("script.sh", exec_entry),
        ],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.file_type("file.txt").unwrap(), FileType::Blob);
    assert_eq!(fs.file_type("link").unwrap(), FileType::Link);
    assert_eq!(fs.file_type("script.sh").unwrap(), FileType::Executable);
}

// ---------------------------------------------------------------------------
// apply — nested paths
// ---------------------------------------------------------------------------

#[test]
fn apply_nested_paths() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    fs.apply(
        &[
            ("a/b/c/deep.txt", WriteEntry::from_text("deep")),
            ("a/b/sibling.txt", WriteEntry::from_text("sibling")),
        ],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("a/b/c/deep.txt").unwrap(), "deep");
    assert_eq!(fs.read_text("a/b/sibling.txt").unwrap(), "sibling");
    assert!(fs.is_dir("a").unwrap());
    assert!(fs.is_dir("a/b").unwrap());
}

// ---------------------------------------------------------------------------
// apply — custom message
// ---------------------------------------------------------------------------

#[test]
fn apply_custom_message() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    fs.apply(
        &[("x.txt", WriteEntry::from_text("x"))],
        &[],
        fs::ApplyOptions {
            message: Some("custom apply msg".into()),
        },
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    let log = fs.log(fs::LogOptions { limit: Some(1), ..Default::default() }).unwrap();
    assert_eq!(log[0].message, "custom apply msg");
}

// ---------------------------------------------------------------------------
// apply — no-op
// ---------------------------------------------------------------------------

#[test]
fn apply_noop_identical_content() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    fs.apply(
        &[("a.txt", WriteEntry::from_text("aaa"))],
        &[],
        Default::default(),
    )
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    let hash1 = fs.commit_hash().unwrap();

    // Same content again
    fs.apply(
        &[("a.txt", WriteEntry::from_text("aaa"))],
        &[],
        Default::default(),
    )
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    let hash2 = fs.commit_hash().unwrap();

    assert_eq!(hash1, hash2);
}

// ---------------------------------------------------------------------------
// apply — stale
// ---------------------------------------------------------------------------

#[test]
fn apply_stale_snapshot_errors() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_old = store.branches().get("main").unwrap();

    // Advance
    let fs_new = store.branches().get("main").unwrap();
    fs_new.write("advance.txt", b"go", Default::default()).unwrap();

    let result = fs_old.apply(
        &[("stale.txt", WriteEntry::from_text("fail"))],
        &[],
        Default::default(),
    );
    assert!(result.is_err());
}

// ---------------------------------------------------------------------------
// apply — empty
// ---------------------------------------------------------------------------

#[test]
fn apply_empty_is_noop() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let hash_before = fs.commit_hash().unwrap();

    fs.apply(&[], &[], Default::default()).unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.commit_hash().unwrap(), hash_before);
}

// ---------------------------------------------------------------------------
// apply — overwrite existing
// ---------------------------------------------------------------------------

#[test]
fn apply_overwrites_existing() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = common::store_with_files(dir.path());

    fs.apply(
        &[("hello.txt", WriteEntry::from_text("overwritten"))],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("hello.txt").unwrap(), "overwritten");
    // Other files preserved
    assert_eq!(fs.read_text("dir/a.txt").unwrap(), "aaa");
}

// ---------------------------------------------------------------------------
// apply — invalid entry validation
// ---------------------------------------------------------------------------

#[test]
fn apply_invalid_entry_errors() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    let bad = WriteEntry {
        data: None,
        target: None,
        mode: MODE_BLOB,
    };
    assert!(fs.apply(&[("bad.txt", bad)], &[], Default::default()).is_err());
}

// ---------------------------------------------------------------------------
// apply — empty data
// ---------------------------------------------------------------------------

#[test]
fn apply_empty_file() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    fs.apply(
        &[("empty.txt", WriteEntry::from_bytes(vec![]))],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read("empty.txt").unwrap(), b"");
    assert_eq!(fs.size("empty.txt").unwrap(), 0);
}

// ---------------------------------------------------------------------------
// apply — binary data
// ---------------------------------------------------------------------------

#[test]
fn apply_binary_data() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    let data: Vec<u8> = (0u8..=255).collect();
    fs.apply(
        &[("binary.bin", WriteEntry::from_bytes(data.clone()))],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read("binary.bin").unwrap(), data);
}

// ---------------------------------------------------------------------------
// apply — preserves other files
// ---------------------------------------------------------------------------

#[test]
fn apply_preserves_existing_files() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = common::store_with_files(dir.path());

    fs.apply(
        &[("new.txt", WriteEntry::from_text("new"))],
        &[],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("hello.txt").unwrap(), "hello");
    assert_eq!(fs.read_text("dir/a.txt").unwrap(), "aaa");
    assert_eq!(fs.read_text("dir/b.txt").unwrap(), "bbb");
    assert_eq!(fs.read_text("new.txt").unwrap(), "new");
}

// ---------------------------------------------------------------------------
// apply — removes
// ---------------------------------------------------------------------------

#[test]
fn apply_removes_single() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = common::store_with_files(dir.path());

    fs.apply(&[], &["hello.txt"], Default::default()).unwrap();

    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("hello.txt").unwrap());
    // Other files preserved
    assert_eq!(fs.read_text("dir/a.txt").unwrap(), "aaa");
}

#[test]
fn apply_removes_multiple() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = common::store_with_files(dir.path());

    fs.apply(&[], &["hello.txt", "dir/a.txt"], Default::default()).unwrap();

    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("hello.txt").unwrap());
    assert!(!fs.exists("dir/a.txt").unwrap());
    assert_eq!(fs.read_text("dir/b.txt").unwrap(), "bbb");
}

#[test]
fn apply_writes_and_removes_combined() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = common::store_with_files(dir.path());

    fs.apply(
        &[("new.txt", WriteEntry::from_text("new"))],
        &["hello.txt"],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("hello.txt").unwrap());
    assert_eq!(fs.read_text("new.txt").unwrap(), "new");
    assert_eq!(fs.read_text("dir/a.txt").unwrap(), "aaa");
}

#[test]
fn apply_writes_and_removes_single_commit() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = common::store_with_files(dir.path());
    let log_before = fs.log(Default::default()).unwrap();

    fs.apply(
        &[("new.txt", WriteEntry::from_text("new"))],
        &["hello.txt"],
        Default::default(),
    )
    .unwrap();

    let fs = store.branches().get("main").unwrap();
    let log_after = fs.log(Default::default()).unwrap();
    assert_eq!(log_after.len(), log_before.len() + 1);
}

#[test]
fn apply_removes_only_is_noop_for_empty() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let hash_before = fs.commit_hash().unwrap();

    // No writes, no removes
    fs.apply(&[], &[], Default::default()).unwrap();

    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.commit_hash().unwrap(), hash_before);
}

File: cpp/tests/test_apply.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <filesystem>
#include <string>
#include <thread>
#include <chrono>

namespace fs = std::filesystem;

static fs::path make_temp_repo() {
    auto tmp = fs::temp_directory_path() /
               ("vost_atest_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_store(const fs::path& path,
                                  const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

// ---------------------------------------------------------------------------
// apply: writes
// ---------------------------------------------------------------------------

TEST_CASE("apply: single write", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"hello.txt", vost::WriteEntry::from_text("world")});
    snap = snap.apply(writes);

    CHECK(snap.read_text("hello.txt") == "world");
    fs::remove_all(path);
}

TEST_CASE("apply: multiple writes", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"a.txt", vost::WriteEntry::from_text("aaa")});
    writes.push_back({"b.txt", vost::WriteEntry::from_text("bbb")});
    writes.push_back({"dir/c.txt", vost::WriteEntry::from_text("ccc")});
    snap = snap.apply(writes);

    CHECK(snap.read_text("a.txt") == "aaa");
    CHECK(snap.read_text("b.txt") == "bbb");
    CHECK(snap.read_text("dir/c.txt") == "ccc");
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: writes + removes
// ---------------------------------------------------------------------------

TEST_CASE("apply: writes and removes combined", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    snap = snap.write_text("old.txt", "old");
    snap = snap.write_text("keep.txt", "keep");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"new.txt", vost::WriteEntry::from_text("new")});

    std::vector<std::string> removes = {"old.txt"};
    snap = snap.apply(writes, removes);

    CHECK(snap.read_text("new.txt") == "new");
    CHECK(snap.read_text("keep.txt") == "keep");
    CHECK_FALSE(snap.exists("old.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: changes report
// ---------------------------------------------------------------------------

TEST_CASE("apply: write new and overwrite existing", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    snap = snap.write_text("existing.txt", "v1");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"existing.txt", vost::WriteEntry::from_text("v2")});
    writes.push_back({"brand_new.txt", vost::WriteEntry::from_text("new")});
    snap = snap.apply(writes);

    CHECK(snap.read_text("existing.txt") == "v2");
    CHECK(snap.read_text("brand_new.txt") == "new");
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: custom message
// ---------------------------------------------------------------------------

TEST_CASE("apply: custom message", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"file.txt", vost::WriteEntry::from_text("data")});

    vost::ApplyOptions opts;
    opts.message = "Custom apply message";
    snap = snap.apply(writes, {}, opts);

    CHECK(snap.message() == "Custom apply message");
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: read-only rejection
// ---------------------------------------------------------------------------

TEST_CASE("apply: read-only Fs throws", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    store.tags().set("v1", snap);
    auto tag_snap = store.tags().get("v1");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"new.txt", vost::WriteEntry::from_text("nope")});
    REQUIRE_THROWS_AS(tag_snap.apply(writes), vost::PermissionError);
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: symlink
// ---------------------------------------------------------------------------

TEST_CASE("apply: symlink entry", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("target.txt", "data");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"link.txt", vost::WriteEntry::symlink("target.txt")});
    snap = snap.apply(writes);

    auto ft = snap.file_type("link.txt");
    CHECK(ft == vost::FileType::Link);
    CHECK(snap.readlink("link.txt") == "target.txt");
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: executable mode via WriteEntry
// ---------------------------------------------------------------------------

TEST_CASE("apply: executable mode via WriteEntry", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    std::string text = "#!/bin/sh";
    std::vector<uint8_t> data(text.begin(), text.end());
    writes.push_back({"script.sh", vost::WriteEntry{std::move(data), std::nullopt, vost::MODE_BLOB_EXEC}});
    snap = snap.apply(writes);

    CHECK(snap.file_type("script.sh") == vost::FileType::Executable);
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: removes
// ---------------------------------------------------------------------------

TEST_CASE("apply: remove single file", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("a.txt", "a");
    snap = snap.write_text("b.txt", "b");

    snap = snap.apply({}, {"a.txt"});
    CHECK_FALSE(snap.exists("a.txt"));
    CHECK(snap.exists("b.txt"));
    fs::remove_all(path);
}

TEST_CASE("apply: remove multiple files", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("a.txt", "a");
    snap = snap.write_text("b.txt", "b");
    snap = snap.write_text("c.txt", "c");

    snap = snap.apply({}, {"a.txt", "b.txt"});
    CHECK_FALSE(snap.exists("a.txt"));
    CHECK_FALSE(snap.exists("b.txt"));
    CHECK(snap.exists("c.txt"));
    fs::remove_all(path);
}

TEST_CASE("apply: null removes are noop", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    // Apply with empty writes and empty removes
    snap = snap.apply({}, {});
    CHECK(snap.exists("file.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: StaleSnapshotError
// ---------------------------------------------------------------------------

TEST_CASE("apply: stale snapshot throws", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap1 = store.branches().get("main");
    auto snap2 = store.branches().get("main");

    // Advance branch via snap1
    snap1.write_text("x.txt", "advance");

    // snap2 is now stale
    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"y.txt", vost::WriteEntry::from_text("data")});
    REQUIRE_THROWS_AS(snap2.apply(writes), vost::StaleSnapshotError);
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: verifies content after add, update, delete
// ---------------------------------------------------------------------------

TEST_CASE("apply: add new file is readable", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"new.txt", vost::WriteEntry::from_text("new")});
    snap = snap.apply(writes);

    CHECK(snap.exists("new.txt"));
    CHECK(snap.read_text("new.txt") == "new");
    fs::remove_all(path);
}

TEST_CASE("apply: update existing file changes content", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "v1");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"file.txt", vost::WriteEntry::from_text("v2")});
    snap = snap.apply(writes);

    CHECK(snap.read_text("file.txt") == "v2");
    fs::remove_all(path);
}

TEST_CASE("apply: delete removes file", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    snap = snap.apply({}, {"file.txt"});

    CHECK_FALSE(snap.exists("file.txt"));
    fs::remove_all(path);
}

TEST_CASE("apply: combined add update delete", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("existing.txt", "old");
    snap = snap.write_text("remove_me.txt", "gone");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"existing.txt", vost::WriteEntry::from_text("updated")});
    writes.push_back({"brand_new.txt", vost::WriteEntry::from_text("new")});
    snap = snap.apply(writes, {"remove_me.txt"});

    CHECK(snap.read_text("existing.txt") == "updated");
    CHECK(snap.read_text("brand_new.txt") == "new");
    CHECK_FALSE(snap.exists("remove_me.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: identical write is noop
// ---------------------------------------------------------------------------

TEST_CASE("apply: identical write preserves tree hash", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");
    auto tree_before = snap.tree_hash();

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"file.txt", vost::WriteEntry::from_text("data")});
    snap = snap.apply(writes);

    CHECK(snap.tree_hash() == tree_before);
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply: operation keyword
// ---------------------------------------------------------------------------

TEST_CASE("apply: operation keyword in message", "[apply]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"file.txt", vost::WriteEntry::from_text("data")});

    vost::ApplyOptions opts;
    opts.operation = "import";
    snap = snap.apply(writes, {}, opts);

    auto msg = snap.message();
    // format_message returns just the operation when no custom message
    CHECK(msg.substr(0, 6) == "import");
    fs::remove_all(path);
}
